File: modules/native-identification-camera/ios/NativeIdentificationCameraModule.swift
import AVFoundation
import ExpoModulesCore

struct NativeCaptureOptions: Record {
  @Field
  var flashMode: String = "auto"
}

struct NativePoint: Record {
  @Field
  var x: Double = 0

  @Field
  var y: Double = 0
}

struct NativeSubjectRegion: Record {
  @Field
  var x: Double = 0

  @Field
  var y: Double = 0

  @Field
  var width: Double = 0

  @Field
  var height: Double = 0
}

public final class NativeIdentificationCameraModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeIdentificationCamera")

    View(NativeIdentificationCameraView.self) {
      Events("onCameraReady", "onCameraError")

      Prop("isActive") { (view: NativeIdentificationCameraView, value: Bool) in
        view.isActive = value
      }

      Prop("position") { (view: NativeIdentificationCameraView, value: String) in
        view.position = value == "front" ? .front : .back
      }

      Prop("maxDetail") { (view: NativeIdentificationCameraView, value: Bool) in
        view.maxDetail = value
      }

      Prop("motionPriority") { (view: NativeIdentificationCameraView, value: Bool) in
        view.motionPriority = value
      }

      Prop("disableLowLightBoost") { (view: NativeIdentificationCameraView, value: Bool) in
        view.disableLowLightBoost = value
      }

      Prop("subjectMetering") { (view: NativeIdentificationCameraView, value: Bool) in
        view.subjectMetering = value
      }

      AsyncFunction("capture") {
        (view: NativeIdentificationCameraView, options: NativeCaptureOptions, promise: Promise) in
        view.capture(options: options, promise: promise)
      }

      AsyncFunction("focus") {
        (view: NativeIdentificationCameraView, point: NativePoint) -> Bool in
        view.focus(normalizedX: point.x, normalizedY: point.y)
      }

      AsyncFunction("setSubjectRegion") {
        (view: NativeIdentificationCameraView, region: NativeSubjectRegion?) -> Bool in
        view.setSubjectRegion(region)
      }
    }
  }
}

File: modules/native-identification-camera/android/src/main/java/expo/modules/nativeidentificationcamera/NativeIdentificationCameraModule.kt
package expo.modules.nativeidentificationcamera

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class NativeCaptureOptions : Record {
  @Field
  var flashMode: String = "auto"
}

class NativePoint : Record {
  @Field
  var x: Double = 0.0

  @Field
  var y: Double = 0.0
}

class NativeSubjectRegion : Record {
  @Field
  var x: Double = 0.0

  @Field
  var y: Double = 0.0

  @Field
  var width: Double = 0.0

  @Field
  var height: Double = 0.0
}

class NativeIdentificationCameraModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NativeIdentificationCamera")

    View(NativeIdentificationCameraView::class) {
      Events("onCameraReady", "onCameraError")

      Prop("isActive") { view: NativeIdentificationCameraView, value: Boolean ->
        view.isActive = value
      }

      Prop("position") { view: NativeIdentificationCameraView, value: String ->
        view.position = value
      }

      Prop("maxDetail") { view: NativeIdentificationCameraView, value: Boolean ->
        view.maxDetail = value
      }

      Prop("motionPriority") { view: NativeIdentificationCameraView, value: Boolean ->
        view.motionPriority = value
      }

      Prop("disableLowLightBoost") { view: NativeIdentificationCameraView, value: Boolean ->
        view.disableLowLightBoost = value
      }

      Prop("subjectMetering") { view: NativeIdentificationCameraView, value: Boolean ->
        view.subjectMetering = value
      }

      AsyncFunction("capture") {
        view: NativeIdentificationCameraView,
        options: NativeCaptureOptions,
        promise: Promise ->
        view.capture(options, promise)
      }

      AsyncFunction("focus") {
        view: NativeIdentificationCameraView,
        point: NativePoint ->
        view.focus(point.x, point.y)
      }

      AsyncFunction("setSubjectRegion") {
        view: NativeIdentificationCameraView,
        region: NativeSubjectRegion? ->
        view.setSubjectRegion(region)
      }
    }
  }
}
